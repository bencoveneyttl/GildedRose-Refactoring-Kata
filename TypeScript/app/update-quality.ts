export type Item = Readonly<{
  name: string;
  sellIn: number;
  quality: number;
}>;

export const enum Names {
  AGED_BRIE = "Aged Brie",
  SULFURAS = "Sulfuras, Hand of Ragnaros",
  BACKSTAGE_PASS = "Backstage passes to a TAFKAL80ETC concert",
  CONJURED = "Conjured Mana Cake",
}

const MAX_QUALITY = 50;

export function updateQuality(items: Readonly<Item[]>): Readonly<Item[]> {
  return items.map((item) => {
    const sellIn = nextSellIn(item);
    const quality = clamp(nextQuality({ ...item, sellIn }), 0, MAX_QUALITY);
    return {
      ...item,
      sellIn,
      quality,
    };
  });
}

function nextSellIn({ name, sellIn }: Item): number {
  switch (name) {
    case Names.SULFURAS:
      return sellIn;
    default:
      return sellIn - 1;
  }
}

function nextQuality({ name, quality, sellIn }: Item): number {
  const expired = sellIn < 0;
  switch (name) {
    case Names.AGED_BRIE:
      return quality + (expired ? 2 : 1);
    case Names.BACKSTAGE_PASS:
      if (expired) {
        return 0;
      } else if (sellIn < 5) {
        return quality + 3;
      } else if (sellIn < 10) {
        return quality + 2;
      }
      return quality + 1;
    case Names.CONJURED:
      return quality - (expired ? 4 : 2);
    case Names.SULFURAS:
      return quality;
    default:
      return quality - (expired ? 2 : 1);
  }
}

function clamp(value: number, min: number, max: number): number {
  return Math.max(Math.min(value, max), min);
}
