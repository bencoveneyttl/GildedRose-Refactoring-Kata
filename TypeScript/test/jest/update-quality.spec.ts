import { Item, Names, updateQuality } from "@/update-quality";

describe("updateQuality", () => {
  it("should handle empty input", () => {
    expect(updateQuality([])).toMatchSnapshot();
  });

  it("should lower SellIn and Quality at the end of each day", () => {
    expect(updateQuality([
      { name: "foo", quality: 5, sellIn: 5 },
      { name: "bar", quality: 10, sellIn: 1 },
      { name: "baz", quality: 1, sellIn: 10 },
    ])).toMatchSnapshot();
  });

  it("should lower quality twice as fast once the sell by date has passed", () => {
    expect(updateQuality([
      { name: "foo", quality: 5, sellIn: 1 },
      { name: "bar", quality: 5, sellIn: 0 },
      { name: "baz", quality: 5, sellIn: -1 },
    ])).toMatchSnapshot();});

  it("should never create a negative quality", () => {
    expect(updateQuality([
      { name: "foo", quality: 0, sellIn: 1 },
      { name: "bar", quality: 1, sellIn: -1 },
      { name: "baz", quality: 0, sellIn: -1 },
    ])).toMatchSnapshot();
  });

  it("should increase the quality for aged brie as time passes", () => {
    expect(updateQuality([
      { name: Names.AGED_BRIE, quality: 5, sellIn: 5 },
      { name: Names.AGED_BRIE, quality: 10, sellIn: 1 },
      { name: Names.AGED_BRIE, quality: 1, sellIn: -1 },
    ])).toMatchSnapshot();
  });

  it("should never increase quality more than 50", () => {
    expect(updateQuality([
      { name: Names.AGED_BRIE, quality: 50, sellIn: 1 },
      { name: Names.AGED_BRIE, quality: 50, sellIn: -1 },
      { name: Names.AGED_BRIE, quality: 49, sellIn: -1 },
    ])).toMatchSnapshot();
  });

  it("should never increase quality more than 50", () => {
    expect(updateQuality([
      { name: Names.AGED_BRIE, quality: 50, sellIn: 1 },
      { name: Names.AGED_BRIE, quality: 50, sellIn: -1 },
      { name: Names.AGED_BRIE, quality: 49, sellIn: -1 },
    ])).toMatchSnapshot();
  });

  it('should never update the sell by date or quality of "Sulfuras"', () => {
    expect(updateQuality([
      { name: Names.SULFURAS, quality: 5, sellIn: 5 },
      { name: Names.SULFURAS, quality: 10, sellIn: 1 },
      { name: Names.SULFURAS, quality: 1, sellIn: 10 },
    ])).toMatchSnapshot();
  });

  it('should increase the quality of "Backstage Passes" by 2 when the sell by date is 10 days or less', () => {
    expect(updateQuality([
      { name: Names.BACKSTAGE_PASS, quality: 10, sellIn: 6 },
      { name: Names.BACKSTAGE_PASS, quality: 10, sellIn: 10 },
      { name: Names.BACKSTAGE_PASS, quality: 10, sellIn: 11 },
    ])).toMatchSnapshot();
  });

  it('should increase the quality of "Backstage Passes" by 3 when the sell by date is 5 days or less', () => {
    expect(updateQuality([
      { name: Names.BACKSTAGE_PASS, quality: 10, sellIn: 5 },
      { name: Names.BACKSTAGE_PASS, quality: 10, sellIn: 3 },
      { name: Names.BACKSTAGE_PASS, quality: 10, sellIn: 1 },
    ])).toMatchSnapshot();
  });

  it('should drop the quality of "Backstage Passes" to 0 when the sell by date has passed', () => {
    expect(updateQuality([
      { name: Names.BACKSTAGE_PASS, quality: 10, sellIn: 0 },
      { name: Names.BACKSTAGE_PASS, quality: 10, sellIn: -1 },
      { name: Names.BACKSTAGE_PASS, quality: 10, sellIn: -2 },
    ])).toMatchSnapshot();
  });

  it('should degrade the quality of "Conjured" items twice as fast', () => {
    expect(updateQuality([
      { name: Names.CONJURED, quality: 5, sellIn: 5 },
      { name: Names.CONJURED, quality: 10, sellIn: 1 },
      { name: Names.CONJURED, quality: 1, sellIn: -1 },
    ])).toMatchSnapshot();
  });
});
