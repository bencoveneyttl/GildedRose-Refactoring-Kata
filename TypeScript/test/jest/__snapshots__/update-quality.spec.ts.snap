// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateQuality should degrade the quality of "Conjured" items twice as fast 1`] = `
Array [
  Object {
    "name": "Conjured Mana Cake",
    "quality": 3,
    "sellIn": 4,
  },
  Object {
    "name": "Conjured Mana Cake",
    "quality": 8,
    "sellIn": 0,
  },
  Object {
    "name": "Conjured Mana Cake",
    "quality": 0,
    "sellIn": -2,
  },
]
`;

exports[`updateQuality should drop the quality of "Backstage Passes" to 0 when the sell by date has passed 1`] = `
Array [
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -1,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -2,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -3,
  },
]
`;

exports[`updateQuality should handle empty input 1`] = `Array []`;

exports[`updateQuality should increase the quality for aged brie as time passes 1`] = `
Array [
  Object {
    "name": "Aged Brie",
    "quality": 6,
    "sellIn": 4,
  },
  Object {
    "name": "Aged Brie",
    "quality": 11,
    "sellIn": 0,
  },
  Object {
    "name": "Aged Brie",
    "quality": 3,
    "sellIn": -2,
  },
]
`;

exports[`updateQuality should increase the quality of "Backstage Passes" by 2 when the sell by date is 10 days or less 1`] = `
Array [
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 12,
    "sellIn": 5,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 12,
    "sellIn": 9,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 11,
    "sellIn": 10,
  },
]
`;

exports[`updateQuality should increase the quality of "Backstage Passes" by 3 when the sell by date is 5 days or less 1`] = `
Array [
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 13,
    "sellIn": 4,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 13,
    "sellIn": 2,
  },
  Object {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 13,
    "sellIn": 0,
  },
]
`;

exports[`updateQuality should lower SellIn and Quality at the end of each day 1`] = `
Array [
  Object {
    "name": "foo",
    "quality": 4,
    "sellIn": 4,
  },
  Object {
    "name": "bar",
    "quality": 9,
    "sellIn": 0,
  },
  Object {
    "name": "baz",
    "quality": 0,
    "sellIn": 9,
  },
]
`;

exports[`updateQuality should lower quality twice as fast once the sell by date has passed 1`] = `
Array [
  Object {
    "name": "foo",
    "quality": 4,
    "sellIn": 0,
  },
  Object {
    "name": "bar",
    "quality": 3,
    "sellIn": -1,
  },
  Object {
    "name": "baz",
    "quality": 3,
    "sellIn": -2,
  },
]
`;

exports[`updateQuality should never create a negative quality 1`] = `
Array [
  Object {
    "name": "foo",
    "quality": 0,
    "sellIn": 0,
  },
  Object {
    "name": "bar",
    "quality": 0,
    "sellIn": -2,
  },
  Object {
    "name": "baz",
    "quality": 0,
    "sellIn": -2,
  },
]
`;

exports[`updateQuality should never increase quality more than 50 1`] = `
Array [
  Object {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": 0,
  },
  Object {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -2,
  },
  Object {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -2,
  },
]
`;

exports[`updateQuality should never increase quality more than 50 2`] = `
Array [
  Object {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": 0,
  },
  Object {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -2,
  },
  Object {
    "name": "Aged Brie",
    "quality": 50,
    "sellIn": -2,
  },
]
`;

exports[`updateQuality should never update the sell by date or quality of "Sulfuras" 1`] = `
Array [
  Object {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 5,
    "sellIn": 5,
  },
  Object {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 10,
    "sellIn": 1,
  },
  Object {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 1,
    "sellIn": 10,
  },
]
`;
